src/test/java/uk/gov/homeoffice/digital/sas/timecard/producers/KafkaProducerServiceTest.java

```diff
@@ -17,60 +17,54 @@
 import org.springframework.test.context.junit.jupiter.SpringExtension;
+import org.springframework.test.util.ReflectionTestUtils;
 import org.springframework.util.concurrent.ListenableFuture;
 import uk.gov.homeoffice.digital.sas.timecard.enums.KafkaAction;
-import uk.gov.homeoffice.digital.sas.timecard.model.KafkaEventMessage;
+import uk.gov.homeoffice.digital.sas.timecard.kafka.KafkaEventMessage;
+import uk.gov.homeoffice.digital.sas.timecard.kafka.producers.KafkaProducerService;
 import uk.gov.homeoffice.digital.sas.timecard.model.TimeEntry;
 
 @ExtendWith(SpringExtension.class)
-class KafkaProducerTimeEntryTest {
+class KafkaProducerServiceTest {
 
   @Mock
-  private KafkaTemplate<String, KafkaEventMessage> kafkaTimeEntryTemplate;
+  private KafkaTemplate<String, KafkaEventMessage<TimeEntry>> kafkaTimeEntryTemplate;
 
   @InjectMocks
-  private KafkaProducerTimeEntry kafkaProducerTimeEntry;
+  private KafkaProducerService<TimeEntry> kafkaProducerService;
 
   @Mock
-  private ListenableFuture<SendResult<String, KafkaEventMessage>> responseFuture;
+  private ListenableFuture<SendResult<String, KafkaEventMessage<TimeEntry>>> responseFuture;
 
   @ParameterizedTest
   @EnumSource(value = KafkaAction.class, names = {"CREATE", "UPDATE"})
   void sendMessage_actionOnTimeEntry_messageIsSentWithCorrectArguments(KafkaAction action) {
-    TimeEntry timeEntry = createTimeEntry();
+    ReflectionTestUtils.setField(kafkaProducerService, "topicName", "callisto-timecard");
+    ReflectionTestUtils.setField(kafkaProducerService, "version", "1.0.0");
+
+    UUID ownerId = UUID.fromString("ec703cac-de76-49c8-b1c4-83da6f8b42ce");
+    LocalDateTime actualStartTime = LocalDateTime.of(
+        2022, 1, 1, 9, 0, 0);
+    TimeEntry timeEntry = createTimeEntry(ownerId, getAsDate(actualStartTime));
 
     Mockito.when(kafkaTimeEntryTemplate.send(Mockito.any(), Mockito.any(), Mockito.any()))
         .thenReturn(responseFuture);
 
     assertThatNoException().isThrownBy(() ->
-        kafkaProducerTimeEntry.sendMessage(timeEntry, action));
+        kafkaProducerService.sendMessage(ownerId.toString(), TimeEntry.class, timeEntry, action));
 
     ArgumentCaptor<String> topicArgument = ArgumentCaptor.forClass(String.class);
     ArgumentCaptor<String> ownerIdArgument = ArgumentCaptor.forClass(String.class);
-    ArgumentCaptor<KafkaEventMessage> messageArgument =
+    ArgumentCaptor<KafkaEventMessage<TimeEntry>> messageArgument =
         ArgumentCaptor.forClass(KafkaEventMessage.class);
 
     Mockito.verify(kafkaTimeEntryTemplate)
         .send(topicArgument.capture(), ownerIdArgument.capture(), messageArgument.capture());
 
     assertThat(topicArgument.getValue()).isEqualTo("callisto-timecard");
     assertThat(ownerIdArgument.getValue()).isEqualTo(timeEntry.getOwnerId().toString());
+    assertThat(messageArgument.getValue().getSchema()).isEqualTo("uk.gov.homeoffice.digital.sas.timecard.model.TimeEntry, 1.0.0");
```

Comment:
Can avoid hardcoding the path/class here with:
`TimeEntry.class.getCanonicalName()`

Reply:
Updated this but not sure it's as easy to understand at a glance. I guess it comes down to readability vs maintainability if the file location ever changed

Reply:
Yeah well put, it's a trade off between the two. My preference is for maintainability but happy either way 👍